src/order/disjointed.lean

```diff
@@ -8,83 +8,139 @@ import order.partial_sups
 /-!
 # Consecutive differences of sets
 
-This file defines a way to make a sequence of sets into a sequence of disjoint sets.
+This file defines the way to make a sequence of elements into a sequence of disjoint elements with
+the same partial sups.
+
+For a sequence `f : ℕ → α`, this new sequence will be `f 0`, `f 1 \ f 0`, `f 2 \ (f 0 ⊔ f 1)`.
+It is actually unique, as
 
 ## Main declarations
 
-* `disjointed f`: Yields the sequence of sets `f 0`, `f 1 \ f 0`, `f 2 \ (f 0 ∪ f 1)`, ... This
-  sequence has the same union as `f 0`, `f 1`, `f 2` but with disjoint sets.
+* `disjointed f`: The sequence `f 0`, `f 1 \ f 0`, `f 2 \ (f 0 ⊔ f 1)`.
+* `partial_sups_disjointed`: `disjointed f` has the same partial sups as `f`
+* `disjoint_disjointed`: The elements of `disjointed f` are pairwise disjoint.
 -/
 
 variables {α β : Type*}
 
-namespace set
+section generalized_boolean_algebra
+variables [generalized_boolean_algebra α]
+
+/-- If `f : ℕ → α` is a sequence of elements, then `disjointed f` is the sequence formed by
+subtracting each element from the nexts. This is the unique disjoint sequence whose partial sups
+are the same as the original sequence. -/
+def disjointed (f : ℕ → α) : ℕ → α
+| 0       := f 0
+| (n + 1) := f (n + 1) \ (partial_sups f n)
 
-/-- If `f : ℕ → set α` is a sequence of sets, then `disjointed f` is
-  the sequence formed with each set subtracted from the later ones
-  in the sequence, to form a disjoint sequence. -/
-def disjointed (f : ℕ → set α) (n : ℕ) : set α := f n ∩ (⋂ i < n, (f i)ᶜ)
+lemma disjointed_zero (f : ℕ → α) : disjointed f 0 = f 0 := rfl
 
-variables {f : ℕ → set α} {n : ℕ}
+lemma disjointed_succ (f : ℕ → α) (n : ℕ) :
+  disjointed f (n + 1) = f (n + 1) \ (partial_sups f n) :=
+rfl
 
-lemma disjoint_disjointed : pairwise (disjoint on disjointed f) :=
-λ i j h, begin
-  wlog h' : i ≤ j; [skip, {revert a, exact (this h.symm).symm}],
-  rintro a ⟨⟨h₁, _⟩, h₂, h₃⟩, simp at h₃,
-  exact h₃ _ (lt_of_le_of_ne h' h) h₁
+lemma disjointed_le_id : disjointed ≤ (id : (ℕ → α) → ℕ → α) :=
+begin
+  rintro f n,
+  cases n,
+  { refl },
+  { exact sdiff_le }
 end
 
--- a more useful version might be `∀ i j x, x ∈ disjointed f i → x ∈ disjointed f j → i = j`
-lemma disjoint_disjointed' :
-  ∀ i j, i ≠ j → (disjointed f i) ∩ (disjointed f j) = ∅ :=
-λ i j hij, disjoint_iff.1 $ disjoint_disjointed i j hij
+lemma disjointed_le (f : ℕ → α) : disjointed f ≤ f := disjointed_le_id f
```

Comment:
I think it makes sense to prove these the other way around, sincethe former is just a weirder way to state the latter.

Reply:
Ah yeah I was about to ask. Do you think `disjointed_le_id` is even useful at all?

Reply:
I think it's probably harmless to have